lib/Socket.js

```diff
@@ -59,71 +51,104 @@ class Socket extends EventEmitter {
     try {
       callback(err, data);
     } catch (e) {
-      log.error('Unable to resolve callback.', e);
+      logAmbiguousError(debug, 'Unable to resolve callback:', e);
+      this.emit('error', e); // Forward the error so that we're not completely swallowing it.
     }
   }
 
-  // TODO: Break this up a bit.
-  // TODO: Clean up callbacks.
-  connect(args = {}, callback) {
-    args = args || {};
-    let address = args.address || 'localhost';
-
-    if (!url.parse(address).port) {
-      address += ':' + DEFAULT_PORT;
-    }
-
-    if (this._connected) {
-      this._socket.close();
-      this._connected = false;
-    }
-
-    return new Promise((resolve, reject) => {
-      log.info('Attempting to connect to:', address);
-      this._socket = new WebSocket('ws://' + address);
-
-      this._socket.onopen = () => {
-        this._connected = true;
-        log.info('Connection opened:', address);
-        this.emit('obs:internal:event', {updateType: 'ConnectionOpened'});
-        this._doCallback(callback, null, true);
-        resolve();
-      };
+  async connect(args = {}, callback) {
+    try {
+      args = args || {};
+      let address = args.address || 'localhost';
 
-      this._socket.onclose = () => {
-        this._connected = false;
-        log.info('Connection closed:', address);
-        this.emit('obs:internal:event', {updateType: 'ConnectionClosed'});
-      };
+      if (!url.parse(address).port) {
+        address += ':' + DEFAULT_PORT;
+      }
 
-      this._socket.onerror = evt => {
+      if (this._connected) {
+        this._socket.close();
         this._connected = false;
-        log.error('Connected failed.', evt.code);
-        this._doCallback(callback, true, null);
-        reject(evt);
-      };
+      }
 
-      this._socket.onmessage = msg => {
-        log.debug('[Socket:OnMessage]', msg);
+      await this._connect(address);
+      this._connected = true;
 
-        const data = camelCaseKeys(JSON.parse(msg.data));
+      // This handler must be present before we can call _authenticate.
+      this._socket.on('message', data => {
+        debug('[OnMessage]: %o', data);
+        data = camelCaseKeys(JSON.parse(data));
 
         // Emit the message with ID if available, otherwise default to a non-messageId driven event.
         if (data.messageId) {
           this.emit('obs:internal:message:id-' + data.messageId, data);
         } else {
           this.emit('obs:internal:event', data);
         }
-      };
-    });
+      });
+
+      await this._authenticate(args.password);
+
+      this._socket.once('close', () => {
```

Comment:
Yeah, just a thought I had since I'm not packing `ws` into the distributable.
Not completely opposed to requiring people run with some enhanced Socket implementation though. Standardizes it across multiple browsers.